components/layout/query.rs

```diff
@@ -44,9 +44,94 @@ use crate::flow::inline::construct::{TextTransformation, WhitespaceCollapse};
 use crate::fragment_tree::{
     BoxFragment, Fragment, FragmentFlags, FragmentTree, SpecificLayoutInfo,
 };
+use crate::style_ext::ComputedValuesExt;
 use crate::taffy::SpecificTaffyGridInfo;
 
-pub fn process_content_box_request<'dom>(node: impl LayoutNode<'dom> + 'dom) -> Option<Rect<Au>> {
+/// Supposed helper for post composite query, query that requires the consideration of transform,
+/// mapping of coordinates, scroll offset, etc.
+pub(crate) struct PostCompositeQueryHelper<'dom> {
+    /// Get the scroll offset of a viewport (i.e. window).
+    pub root_scroll_offset_getter: &'dom dyn Fn() -> Vector2D<Au>,
+    /// Get the scroll offset of a BoxFragment.
+    pub scroll_offset_getter: &'dom dyn Fn(&BoxFragment) -> Vector2D<Au>,
+}
+
+impl PostCompositeQueryHelper<'_> {
+    fn retrieve_box_fragment(fragment: &Fragment) -> Option<&ArcRefCell<BoxFragment>> {
+        match fragment {
+            Fragment::Box(box_fragment) | Fragment::Float(box_fragment) => Some(box_fragment),
+            _ => None,
+        }
+    }
+
+    /// Get the position property for current node.
+    fn retrieve_position_property<'dom, Node: LayoutNode<'dom> + 'dom>(
+        node: Node,
+    ) -> Option<Position> {
+        node.fragments_for_pseudo(None)
+            .first()
+            .and_then(PostCompositeQueryHelper::retrieve_box_fragment)
+            .map(|fragment| fragment.borrow().style.get_box().position)
+    }
+
+    /// Find the parent node in containing block chain.
+    fn parent_containing_block<'dom, Node: LayoutNode<'dom> + 'dom>(
+        node: Node,
+        position: Option<Position>,
+    ) -> Option<Node> {
+        fn established_containing_block(
+            style: &ServoArc<ComputedValues>,
+            fragment_flags: FragmentFlags,
+            position: Option<Position>,
+        ) -> bool {
+            match position {
+                Some(Position::Absolute) => {
+                    style.establishes_containing_block_for_absolute_descendants(fragment_flags)
+                },
+                Some(Position::Fixed) => {
+                    style.establishes_containing_block_for_all_descendants(fragment_flags)
+                },
```

Comment:
```suggestion
                Some(Position::Absolute) | Some(Position::Fixed) => {
                    style.establishes_containing_block_for_absolute_descendants(fragment_flags)
                },
```
This is because `fixed` is just a special case for `absolute`, and `establishes_containing_block_for_all_descendants` implies `establishes_containing_block_for_absolute_descendants`

Reply:
I am not sure, I suppose in containing block terms, `fixed` and `absolute` need to be differentiated, since there could be a box that establishes a containing block for `absolute` but not for `fixed` (e.g. box with `position: relative`).

Reply:
I see, you are right.